### DISCUSSION TITLE
Strategy : NFT-DAO Test Strategy

### DISCUSSION THREAD

**Discussion by stephen-rowan** [NFT-DAO/Governance-HOLON#76] [Strategy]
# NFT-DAO Test Strategy

Hey all - a few weeks ago I think Philip drew attention to a failed NFT launch and the associated reputation damage it caused. 
At that time, I raised the possibility of treating the first bunch of NFTs as a test programme - "Come break our tokens" etc. 
To that end, I have drafted a rough outline for an NFT-DAO Test Strategy. I would appreciate any feedback.

## Scope

**This is post-development testing**. Everything from User Acceptance Testing (UAT) to iterative testing of NFT-DAO processes, procedures and methods. 

## Challenges

The challenge is how to adapt quality and maturity levels to a distributed, autonomous approach. One possible route is to start with a Test Strategy - where all NFT-DAO's processes, procedures and methods are tested in the community.

A PDF of a outline document is at https://github.com/NFT-DAO/Governance-HOLON/blob/main/Documents/Testing/NFT-DAO-Test-Stategy.pdf

### Short-Term : 

Treat one of the MVP Releases as a User Acceptance Test (UAT) - to achieve a "Defined" maturity level.
Identify Token Risk Areas.
Run Test Cases with Subscribers
Document outputs to demonstrate NFT-DAO can achieve  a "Defined" maturity level.

### Medium-Term : 

NFT-DAO adopts some kind of quality testing method - to achieve a "Integration" maturity level.

### Long-Term : 

NFT-DAO's whole approach is a test lifecycle - to achieve a "Optimization" maturity level.

## Blockchain Test Maturity Model

This is an adaption of KPMG’s Blockchain Maturity Model oriented towards Test Maturity Models.

The challenge is adapting maturity levels to distributed autonomous organisations (DAO). 

### Level 1 – Initial

Ad-hoc testing.

### Level 2 – Defined.

Testing does not start until products are completed, so the aim of testing is to compare products against requirements.

### Level 3 – Integration

Testing is integrated into a quality life cycle,

### Level 4 – Management and measurement

Testing activities take place at all stages of the life cycle.

### Level 5 – Optimization

Testing process itself is tested and improved at each iteration.

**Comment by stephen-rowan** [comment]
# Risk Areas

Possible risk areas NFT-DAO may need to test. (Adapted from KPMG's Blockchain Maturity Model)

## 1. Access and user management

1.1 Management of cryptographic keys.

Transparency of policy etc

1.2 Unauthorized access of participants

How to handle unauthorised or fraudulent access. 
Cardano Registry has a legal team looking at this in their area.

1.3 How to uniquely identify users.

DID to avoid duplicate identities in shared auctions etc.

## 2. Authorization and provisioning management

2.1 Segregation of duties

Due diligence in assigning responsibility to stakeholders in NFT minting.

2.2 Incorrect authorizations

How to handle errors

2.3 Abuse of high privileged or over authorized users.

How to play fairly
 
## 3. Data management

Data confidentiality

Data integrity

Data availability

## 4. Interoperability

Integrating with legacy systems

For example, off-chain, web 3 and GitHub registry

## 5. Scalability and performance

Repeats earlier issues above in 1 and 2 (management of cryptographic keys, unauthorized access of participants and uniquely identifiable users) at scale.

## 6. Change management.

HOLON agreement on handover between tasks by all participants

Standards adoption in each HOLON

GitHub Version control: forking

## 7. Privacy

Append-only data structure (?)

The ‘right to be forgotten’.

GDPR. (European Privacy Legislation)

## 8. Security

Coherence of the consensus mechanism chosen.

the number and location of nodes

**Reply by Yagi-Michael** [reply]
This is a great effort! What would be the best way to run do this? Having a clearly defined checklist would be great. It will probably evolve over time what needs to be in it, but we could "dogfood" and use the MVP as a test run for this.

**Reply by stephen-rowan** [reply]
Great to have your support @Yagi-Michael - I am putting together a project board over at https://github.com/NFT-DAO/NFT-Standards/projects/2 (NFT-Standards is probably a more appropriate place for a Test Strategy ?). I will move everything over there if you agree ?

I have Test Management experience - but really I am hoping we can innovate together and create our own new DAO approach. Mix and match our different backgrounds. As you suggest start simple with an MVP test run, develop and iterate over time. 

I will start to put together a Test Strategy Document in markdown on NFT-Standards (so we can do pull requests on it and track it's development) and work out how to co-ordinate all the HOLONs via the Project Board. 

An overview of some of the things I have in mind (step in and comment) - 

- [ ] Add a Testing channel to NFT-DAO Slack to get Developer involvement from the outset.
- [ ] Identify what test platforms will available for different MVP launches.
- [ ] Work out how to capture Test Cases for each participant on different platforms/offerings. 
- [ ] Identify what Risk Areas we are most vulnerable to and prioritize their UAT.
- [ ] Agree a Test Timetable, with handovers between HOLONs and support in place etc.
- [ ] Market and Communicate the test as an offering in itself "Come and Break our Token" etc. (to incentive involvement).
- [ ] Work out if we can actually tokenise incentives to take part at this stage (e.g. gamify the test).
- [ ] Document the whole thing and present as evidence that NFT-DAO can achieve Maturity level 2.
- [ ] I am going to do assessment of the failures of the OccamFi launch and see how it fits into KPMG's Blockchain Risk Model

Have a good weekend and hopefully we can catch up sometime next week.

S

**Reply by Yagi-Michael** [reply]
I was thinking more along the lines of a checklist for completion. For example, test coverage for scenarios xyz, support integration with abc, that kind of thing. The above would be working towards what I was talking about, but referencing some other audit processes might be a good starting point

**Reply by stephen-rowan** [reply]
Yes, "_test coverage for scenarios xyz, support integration with abc_" would be covered in the drafting of Test Cases from a UAT perspective. I was going to concentrate on KPMG's risk areas ( https://github.com/NFT-DAO/NFT-Standards/blob/main/Testing/Documentation/02-Risks.md ) and build test cases for those. That is already quite a lot of work.

For example one checklist item (from 1. Access and user management) for a test case would break down - 

> 1.2 Unauthorized access of participants
> How to handle unauthorized or fraudulent access. 

into a step-by-step scenario (we can setup on a test platform) and track. We would need to put in place a procedure on how we would handle such a scenario and test it.

For User testing - integration would take the form of a test workflow. We would build a list of things we want the User to test and in what order etc.

S
